### DISCUSSION TITLE
Fallback error page as a Svelte component

### DISCUSSION THREAD

**Discussion by coryvirok** [sveltejs/kit#8393] [Ideas]
In order to do some advanced navigation and auth guards I find myself having to put more and more logic into `hooks.server.ts`. And since errors in this part of the request lifecycle are rendered using the fallback error page, it's not possible to customize that error page very easily.

Use case: auth guard where going to a protected route while not logged in should always render the login page/component **at that URL**.

1. Navigate to `/protected`
2. `hooks.server.ts#handle()` notices the route is protected and runs `throw error(401)`
3. fallback error page is rendered (no redirects - it's rendered at the `/protected` url)
   - It checks `[401, 402].includes(statusCode)` and renderes the `Login` component
   - Otherwise, renders an error page

The reason I'd like to have the fallback error page be a Svelte component (rather than the current "dumb" template) is because I'd like to customize the error page based on status code. As well as bring in any compiled CSS or other assets.

An alternative solution would be to enable different static fallback error pages based on status code like some other frameworks allow. e.g. 401.html, 404.html, etc.

**Comment by pevey** [comment]
You could handle this in the layout page. The layout gets rendered along with the error page. Or even just within hooks itself you can do the check for instance if the route is /protected and the user is not authenticated, and then throw a redirect to the login page instead of throwing a 401 error.

**Reply by coryvirok** [reply]
Layouts don't get rendered if there is an error thrown in the handle hook. Also, I'd like to not redirect to /login but rather render the login page at the url that requires auth. 

Putting any auth checks into a layout is dangerous. See #6315

**Comment by 7nik** [comment]
What is wrong with redirecting to the login page?
As for me, the error.html is for fatal errors when uncaught exception happened and thus the server unable to generate an adequate response.
Your case is completely handled and server can return whatever you want.

If the error.html will be a component, what should the server respond if error occur during the component rendering? It already tries to render the *fallback* error page. This is why the error.html must be as simple as possible.

**Reply by wkille** [reply]
I understand the need for error.html but the OP (and I) would like a more easily customisable error page for _expected_ errors thrown in hooks. I don't like redirecting with a 3xx code when the request deserves a 4xx response.

**Comment by devtrix** [comment]
I agree with OP & @wkille .  We need a better solution.  Also I feel I have to do quite a bit of gymnastics to style the error.html.  I chose skeleton.dev with tailwindcss.  It looks like many of the directives like, %sveltekit.assets%/ doesnt work.  I am going to create a new discussion and reference this.

**Reply by oluijks** [reply]
And I agree with you. People want to style error pages and the fallback. There should be some mechanism to include components or have some layout available to include some styling. Is there something we are missing that makes this nearly impossible?

**Comment by ayazemre** [comment]
I would like to have this option as well. It would be easier if I was building my fallback page as .svelte file instead of barebone .html. I use this workaround at the moment to redirect my own error page and show the page content's based on the query details.

```
<!DOCTYPE html>
<html lang="en">

<head>
	<meta charset="utf-8" />
</head>

<body>
	<script>
		window.location = `/error?status=%sveltekit.status%&message=%sveltekit.error.message%`
	</script>
</body>

</html>
```
